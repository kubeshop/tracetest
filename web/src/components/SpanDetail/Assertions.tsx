import {useState} from 'react';
import {Typography} from 'antd';
import {PlusOutlined} from '@ant-design/icons';
import {ISpan} from 'types/Span.types';
import SkeletonTable from 'components/SkeletonTable';
import AssertionsResultTable from '../AssertionsTable';
import * as S from './SpanDetail.styled';
import CreateAssertionModal from '../CreateAssertionModal';
import {IAssertionResultList} from '../../types/Assertion.types';

interface IProps {
  span?: ISpan;
  testId?: string;
  resultId?: string;
  assertionsResultList: IAssertionResultList[];
}

const Assertion: React.FC<IProps> = ({assertionsResultList, testId, span, resultId}) => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  return (
    <SkeletonTable loading={!span}>
      <S.AssertionActionsContainer>
        <S.AddAssertionButton
          data-cy="add-assertion-button"
          icon={<PlusOutlined />}
          onClick={() => setIsModalOpen(true)}
        >
          Add Assertion
        </S.AddAssertionButton>
      </S.AssertionActionsContainer>
      {assertionsResultList.length ? (
        assertionsResultList
          .sort((a, b) => (a.assertion.assertionId > b.assertion.assertionId ? -1 : 1))
          .map(({assertion, assertionResultList}, index) => (
            <AssertionsResultTable
              key={assertion.assertionId}
              assertionResults={assertionResultList}
              sort={index + 1}
              assertion={assertion}
              span={span!}
              resultId={resultId!}
              testId={testId!}
            />
          ))
      ) : (
        <S.DetailsEmptyStateContainer data-cy="empty-assertion-table">
          <S.DetailsTableEmptyStateIcon />
          <Typography.Text disabled>No Data</Typography.Text>
        </S.DetailsEmptyStateContainer>
      )}

      <CreateAssertionModal
        open={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        span={span!}
        testId={testId!}
        resultId={resultId!}
      />
    </SkeletonTable>
  );
};

export default Assertion;
