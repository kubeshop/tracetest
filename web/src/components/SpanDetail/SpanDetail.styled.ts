import {Divider, Typography} from 'antd';
import styled from 'styled-components';

export const Header = styled.div`
  padding: 16px 12px 0;
  min-width: 270px;
`;

export const HeaderCheck = styled(Typography.Text)`
  color: ${({theme}) => theme.color.text};
  font-size: ${({theme}) => theme.size.sm};
  margin-right: 8px;
`;

export const HeaderDot = styled.span<{$passed: boolean}>`
  background-color: ${({$passed, theme}) => ($passed ? theme.color.success : theme.color.error)};
  height: 10px;
  width: 10px;
  display: inline-block;
  margin-right: 4px;
  line-height: 0;
  vertical-align: -0.1em;
  border-radius: 50%;
`;

export const HeaderDivider = styled(Divider)`
  margin: 16px 0;
`;

export const HeaderItem = styled.div`
  align-items: center;
  color: ${({theme}) => theme.color.text};
  display: flex;
  font-size: ${({theme}) => theme.size.md};
  margin-right: 8px;
`;

export const HeaderItemText = styled(Typography.Text)`
  color: inherit;
  margin-left: 5px;
`;

export const HeaderTitle = styled(Typography.Title)`
  && {
    margin: 0;
  }
`;

export const Column = styled.div`
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  margin-bottom: 4px;
`;

export const Row = styled.div`
  align-items: center;
  display: flex;
  margin-bottom: 4px;
`;

export const AttributesContainer = styled.div<{$top: number}>`
  height: calc(100% - ${({$top}) => `${$top}px`});
  overflow-y: scroll;
`;

export const SearchContainer = styled.div`
  padding: 0 12px;
`;
