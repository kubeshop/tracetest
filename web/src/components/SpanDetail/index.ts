import TestAttributeRow from './TestAttributeRow/TestAttributeRow';
import TestSubHeader from './TestSubHeader/TestSubHeader';
import TraceAttributeRow from './TraceAttributeRow/TraceAttributeRow';
import TraceSubHeader from './TraceSubHeader/TraceSubHeader';

// eslint-disable-next-line no-restricted-exports
export {default} from './SpanDetail';
export {TestAttributeRow, TestSubHeader, TraceAttributeRow, TraceSubHeader};
