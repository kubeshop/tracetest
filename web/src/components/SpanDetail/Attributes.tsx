import {useEffect, useRef, useState} from 'react';
import AttributeList from 'components/AttributeList';
import {OtelReference} from 'components/TestSpecForm/hooks/useGetOTELSemanticConventionAttributesInfo';
import TestRunOutput from 'models/TestRunOutput.model';
import {TSpanFlatAttribute} from 'types/Span.types';
import {TAnalyzerError, TTestSpecSummary} from 'types/TestRun.types';
import {IPropsAttributeRow} from './SpanDetail';
import * as S from './SpanDetail.styled';

interface IProps {
  attributeList: TSpanFlatAttribute[];
  searchText?: string;
  semanticConventions: OtelReference;
  analyzerErrors?: TAnalyzerError[];
  testSpecs?: TTestSpecSummary;
  testOutputs?: TestRunOutput[];
  onCreateTestSpec(attribute: TSpanFlatAttribute): void;
  onCreateOutput(attribute: TSpanFlatAttribute): void;
  AttributeRowComponent: React.ComponentType<IPropsAttributeRow>;
}

const Attributes = ({
  attributeList,
  searchText,
  semanticConventions,
  analyzerErrors,
  testSpecs,
  testOutputs,
  onCreateTestSpec,
  onCreateOutput,
  AttributeRowComponent,
}: IProps) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const [topPosition, setTopPosition] = useState(0);

  useEffect(() => {
    setTopPosition(containerRef?.current?.offsetTop ?? 0);
  }, [attributeList]);

  return (
    <S.AttributesContainer $top={topPosition} ref={containerRef}>
      <AttributeList
        attributeList={attributeList}
        searchText={searchText}
        semanticConventions={semanticConventions}
        analyzerErrors={analyzerErrors}
        testSpecs={testSpecs}
        testOutputs={testOutputs}
        onCreateTestSpec={onCreateTestSpec}
        onCreateOutput={onCreateOutput}
        AttributeRowComponent={AttributeRowComponent}
      />
    </S.AttributesContainer>
  );
};

export default Attributes;
