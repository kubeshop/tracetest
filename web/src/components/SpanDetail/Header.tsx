import {SettingOutlined, ToolOutlined} from '@ant-design/icons';
import {Space} from 'antd';
import * as SSpanNode from 'components/Visualization/components/DAG/BaseSpanNode/BaseSpanNode.styled';
import {SemanticGroupNamesToText} from 'constants/SemanticGroupNames.constants';
import Span from 'models/Span.model';
import SpanService from 'services/Span.service';
import * as S from './SpanDetail.styled';
import {StepsID} from '../GuidedTour/testRunSteps';

interface IProps {
  span?: Span;
}

const Header = ({span}: IProps) => {
  const {kind, name, service, system, type} = SpanService.getSpanInfo(span);

  if (!span) {
    return (
      <S.Header data-tour={StepsID.SpanDetails}>
        <S.HeaderTitle level={3}>Span Attributes</S.HeaderTitle>
      </S.Header>
    );
  }

  return (
    <S.Header data-tour={StepsID.SpanDetails}>
      <S.Column>
        <Space>
          <SSpanNode.BadgeType $hasMargin count={SemanticGroupNamesToText[type]} $type={type} />
        </Space>
        <S.HeaderTitle level={3}>{name}</S.HeaderTitle>
      </S.Column>
      <S.Column>
        <S.HeaderItem>
          <SettingOutlined />
          <S.HeaderItemText>{`${service} ${kind}`}</S.HeaderItemText>
        </S.HeaderItem>
        {Boolean(system) && (
          <S.HeaderItem>
            <ToolOutlined />
            <S.HeaderItemText>{system}</S.HeaderItemText>
          </S.HeaderItem>
        )}
      </S.Column>
    </S.Header>
  );
};

export default Header;
