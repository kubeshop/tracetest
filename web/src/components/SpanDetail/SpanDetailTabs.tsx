import {Tabs} from 'antd';
import {capitalize} from 'lodash';
import React, {useMemo} from 'react';
import SpanAttributeService from 'services/SpanAttribute.service';
import AttributeList from '../AttributeList';
import {TSpanDetailsComponentProps} from './SpanDetail';
import * as S from './SpanDetail.styled';

const SpanDetailTabs: React.FC<TSpanDetailsComponentProps> = ({span: {attributeList = [], type} = {}, onCreateAssertion}) => {
  const sectionList = useMemo(
    () => SpanAttributeService.getSpanAttributeSectionsList(attributeList, type!),
    [attributeList, type]
  );

  return (
    <S.SpanTabs data-cy="span-details-attributes">
      {sectionList.map(({section, attributeList: attrList}) => (
        <Tabs.TabPane tab={capitalize(section)} key={section}>
          <AttributeList attributeList={attrList} onCreateAssertion={onCreateAssertion} />
        </Tabs.TabPane>
      ))}
    </S.SpanTabs>
  );
};

export default SpanDetailTabs;
