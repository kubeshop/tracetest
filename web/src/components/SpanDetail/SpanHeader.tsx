import {Typography} from 'antd';
import GuidedTourService, {GuidedTours} from '../../services/GuidedTour.service';
import {Steps} from '../GuidedTour/traceStepList';
import * as S from './SpanDetail.styled';

interface IProps {
  title: string;
}

const SpanHeader: React.FC<IProps> = ({title}) => (
  <S.SpanHeader data-tour={GuidedTourService.getStep(GuidedTours.Trace, Steps.Details)}>
    <S.SpanHeaderTitle>Span Details</S.SpanHeaderTitle>
    <Typography.Text type="secondary">{title}</Typography.Text>
  </S.SpanHeader>
);

export default SpanHeader;
