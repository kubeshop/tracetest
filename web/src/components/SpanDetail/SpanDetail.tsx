import {capitalize} from 'lodash';
import {useCallback} from 'react';
import {SemanticGroupNamesToText} from '../../constants/SemanticGroupNames.constants';
import SpanService from '../../services/Span.service';
import {TSpan, TSpanFlatAttribute} from '../../types/Span.types';
import SpanHeader from './SpanHeader';
import * as S from './SpanDetail.styled';
import {useAssertionForm} from '../AssertionForm/AssertionFormProvider';
import {CompareOperator} from '../../constants/Operator.constants';
import OperatorService from '../../services/Operator.service';
import SpanDetailTabs from './SpanDetailTabs';

export interface TSpanDetailProps {
  testId?: string;
  span?: TSpan;
  resultId?: string;
}

export interface TSpanDetailsComponentProps {
  span?: TSpan;
  onCreateAssertion(attribute: TSpanFlatAttribute): void;
}

const getSpanTitle = (span: TSpan) => {
  const {primary, heading} = SpanService.getSpanNodeInfo(span);
  const spanTypeText = SemanticGroupNamesToText[span.type];

  return `${capitalize(heading) || spanTypeText} • ${primary} • ${span.name}`;
};

const SpanDetail: React.FC<TSpanDetailProps> = ({span}) => {
  const {open} = useAssertionForm();

  const title = (span && getSpanTitle(span)) || '';

  const onCreateAssertion = useCallback(
    ({value, key}: TSpanFlatAttribute) => {
      const {selectorList, pseudoSelector} = SpanService.getSelectorInformation(span!);

      open({
        isEditing: false,
        defaultValues: {
          pseudoSelector,
          assertionList: [
            {
              comparator: OperatorService.getOperatorSymbol(CompareOperator.EQUALS),
              expected: value,
              attribute: key,
            },
          ],
          selectorList,
        },
      });
    },
    [open, span]
  );

  return (
    <S.SpanDetail>
      <SpanHeader title={title} />
      <SpanDetailTabs onCreateAssertion={onCreateAssertion} span={span} />
    </S.SpanDetail>
  );
};

export default SpanDetail;
