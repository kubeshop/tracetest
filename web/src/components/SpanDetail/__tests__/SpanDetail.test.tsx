import faker from '@faker-js/faker';
import {render} from '@testing-library/react';
import SpanMock from '../../../models/__mocks__/Span.mock';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import SpanDetail from '../SpanDetail';

test('Layout', () => {
  const result = render(
    <ReduxWrapperProvider>
      <SpanDetail resultId={faker.datatype.uuid()} testId={faker.datatype.uuid()} span={SpanMock.model()} />
    </ReduxWrapperProvider>
  );
  expect(result.container).toMatchSnapshot();
});
