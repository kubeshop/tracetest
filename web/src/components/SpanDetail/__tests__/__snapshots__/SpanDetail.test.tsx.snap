// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout 1`] = `
<div>
  <div
    class="sc-gKclnd bwOPWE"
  >
    <div
      class="sc-bdvvtL NRNkx"
    >
      <h5
        class="ant-typography sc-gsDKAQ cksOeG"
      >
        Span Details
      </h5>
      <span
        class="ant-typography ant-typography-secondary"
      >
        General • span-mock • span-mock
      </span>
    </div>
    <div
      class="ant-tabs ant-tabs-top sc-dkPtRN gPnAnM"
      data-cy="span-details-attributes"
      style="margin-bottom: 150px; overflow: hidden;"
    >
      <div
        class="ant-tabs-nav"
        role="tablist"
      >
        <div
          class="ant-tabs-nav-wrap"
        >
          <div
            class="ant-tabs-nav-list"
            style="transform: translate(0px, 0px);"
          >
            <div
              class="ant-tabs-tab ant-tabs-tab-active"
            >
              <div
                aria-controls="rc-tabs-test-panel-custom"
                aria-selected="true"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-custom"
                role="tab"
                tabindex="0"
              >
                Custom
              </div>
            </div>
            <div
              class="ant-tabs-tab"
            >
              <div
                aria-controls="rc-tabs-test-panel-all"
                aria-selected="false"
                class="ant-tabs-tab-btn"
                id="rc-tabs-test-tab-all"
                role="tab"
                tabindex="0"
              >
                All
              </div>
            </div>
            <div
              class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
            />
          </div>
        </div>
        <div
          class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
        >
          <button
            aria-controls="rc-tabs-test-more-popup"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-hidden="true"
            class="ant-tabs-nav-more"
            id="rc-tabs-test-more"
            style="visibility: hidden; order: 1;"
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="ellipsis"
              class="anticon anticon-ellipsis"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="ellipsis"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="ant-tabs-content-holder"
      >
        <div
          class="ant-tabs-content ant-tabs-content-top"
        >
          <div
            aria-hidden="false"
            aria-labelledby="rc-tabs-test-tab-custom"
            class="ant-tabs-tabpane ant-tabs-tabpane-active"
            id="rc-tabs-test-panel-custom"
            role="tabpanel"
            tabindex="0"
          >
            <div
              class="sc-fotOHu hXnbHw"
              data-cy="empty-attribute-list"
            >
              <span
                aria-label="question-circle"
                class="anticon anticon-question-circle sc-fFeiMQ rFEFY"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="question-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                  />
                </svg>
              </span>
              <h5
                class="ant-typography ant-typography-secondary sc-bkkeKt"
              >
                Looking for attributes?
              </h5>
              <span
                class="ant-typography ant-typography-secondary sc-ieecCq lgsuex"
              >
                Take a look a the open telemetry specification for
                 
                <a
                  href="https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions"
                >
                  trace semantic conventions
                </a>
                 and here for
                 
                <a
                  href="https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/resource/semantic_conventions"
                >
                  resource semantic conventions
                </a>
                .
              </span>
            </div>
          </div>
          <div
            aria-hidden="true"
            aria-labelledby="rc-tabs-test-tab-all"
            class="ant-tabs-tabpane"
            id="rc-tabs-test-panel-all"
            role="tabpanel"
            style="display: none;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
